### PR TITLE
rustbuild: Support specifying archiver and linker explicitly

### PR DESCRIPTION
**Pull request by petrochenkov**
With this patch `x.py test` passes without toolchain being in `PATH` if `cc`, `cxx`, `ar`, `linker` and `gdb` are specified in `config.toml` (except for a few `run-make` tests using `nm`).

Fixes https://github.com/rust-lang/rust/issues/41821
r? @Mark-Simulacrum

### CODE REVIEW

File: src/bootstrap/builder.rs [hunk 1]
```diff
@@ -413,12 +413,14 @@ impl<'a> Builder<'a> {
...
-            .env("RUSTC_LIBDIR", self.sysroot_libdir(compiler, self.build.build))
-            .env("CFG_RELEASE_CHANNEL", &self.build.config.channel)
-            .env("RUSTDOC_REAL", self.rustdoc(host));
+        cmd.env("RUSTC_STAGE", compiler.stage.to_string())
+           .env("RUSTC_SYSROOT", self.sysroot(compiler))
+           .env("RUSTC_LIBDIR", self.sysroot_libdir(compiler, self.build.build))
+           .env("CFG_RELEASE_CHANNEL", &self.build.config.channel)
+           .env("RUSTDOC_REAL", self.rustdoc(host));
+        if let Some(linker) = self.build.linker(host) {
+            cmd.arg("--linker").arg(linker).arg("-Z").arg("unstable-options");
```
Comment (Mark-Simulacrum) [inline src/bootstrap/builder.rs]: I'd prefer this logic to be in `src/bin/rustdoc.rs` and we can just set `RUSTODOC_LINKER` here.

**Review by Mark-Simulacrum — APPROVED** [review]
r=me pending review of the base PR, at least in terms of logic seeming sound. Not familiar enough with linkers and such to judge that part, but if it works, I'm happy with it.

File: src/bootstrap/builder.rs (hunk 1 above)
Comment (petrochenkov) [reply src/bootstrap/builder.rs]: Hm, this could actually solve the issue with doctests as well if I understand correctly.

**Comment by petrochenkov** [comment]
To give some background, I need to build Rust on a Linux server 1) using specific toolchain that is not in `PATH` 2) that toolchain has `c++`, but doesn't have `cc` (only `gcc`) and doesn't have `ar` (only `gar` and `gcc-ar` wrapper) 3) there's another toolchain in `PATH`, but it's ancient and unusable.

Without this patch my workaround looks roughly like this (in shell pseudocode):
```sh
mkdir my_toolchain
ln -s actual_toolchain/gcc my_toolchain/cc
ln -s actual_toolchain/gar my_toolchain/ar
export PATH=my_toolchain:actual_toolchain:$PATH
./x.py build
```
Even ignoring `ar`, setting `export CC=gcc` (or `cc = gcc` in `config.toml`) is not enough because `rustc` will still attempt to use `cc` for linking Rust code and fail.

**Comment by kennytm** [comment]
Failed to find the symbol `res_init` in `dist-x86_64-musl` when building stage2-std (gnu → musl), legit. (needs to link to `libresolv`?)

<details>

```
[00:42:45] error: linking with `/musl-x86_64/bin/musl-gcc` failed: exit code: 1
[00:42:45]   |
[00:42:45]   = note: "/musl-x86_64/bin/musl-gcc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-std/release/build/unwind-4fb8ab0cb26a92ad/build_script_build-4fb8ab0cb26a92ad.0.o" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-std/release/build/unwind-4fb8ab0cb26a92ad/build_script_build-4fb8ab0cb26a92ad" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-std/release/build/unwind-4fb8ab0cb26a92ad/build_script_build-4fb8ab0cb26a92ad.crate.allocator.o" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-std/release/deps" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-ee965bf30dd84348.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/librand-11d028a4b5a1ee6a.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-2e20520a7c184bee.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_system-ea2f43e1e71cad7c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-bb8002df75b7c03d.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-d5bdce442262816c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-74bec6c5a1a4a26e.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-dbb735482143346f.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_unicode-646edc526b22cddf.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-beef29bda9c153db.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-1a1fb4116f61f74c.rlib" "-Wl,-Bdynamic" "-l" "dl" "-l" "rt" "-l" "pthread" "-l" "pthread" "-l" "gcc_s" "-l" "c" "-l" "m" "-l" "rt" "-l" "pthread" "-l" "util"
[00:42:45]   = note: /usr/bin/ld: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-ee965bf30dd84348.rlib(std-ee965bf30dd84348.0.o): undefined reference to symbol '__res_init@@GLIBC_2.2.5'
[00:42:45]           //lib/x86_64-linux-gnu/libc.so.6: error adding symbols: DSO missing from command line
[00:42:45]           collect2: error: ld returned 1 exit status
[00:42:45]           
[00:42:45] 
[00:42:45] error: aborting due to previous error
[00:42:45] 
[00:42:45] error: Could not compile `unwind`.
```

</details>

**Comment by petrochenkov** [comment]
Oops, that was silly.
@bors r=Mark-Simulacrum

**Pull request merged**